File: src/algorithms/arrays/sliding-window/subarray-product-less-than-k/sources/subarray-product-less-than-k.py
# Subarray Product < K — O(n) variable sliding window counting subarrays with product below threshold
def subarray_product_less_than_k(input_array: list[int], threshold: int) -> dict:
    if len(input_array) == 0 or threshold <= 1:  # @step:initialize
        return {"count": 0}  # @step:initialize

    left_pointer = 0  # @step:initialize
    current_product = 1
    count = 0

    # Expand the right boundary of the window
    for right_pointer in range(len(input_array)):
        current_product *= input_array[right_pointer]  # @step:expand-window

        # Shrink from the left while product meets or exceeds threshold
        while current_product >= threshold and left_pointer <= right_pointer:  # @step:compare
            current_product //= input_array[left_pointer]  # @step:shrink-window
            left_pointer += 1  # @step:shrink-window

        # Every subarray ending at right_pointer and starting anywhere in [left_pointer, right_pointer]
        count += right_pointer - left_pointer + 1  # @step:compare

    return {"count": count}  # @step:complete

File: src/algorithms/arrays/sliding-window/subarray-product-less-than-k/sources/SubarrayProductLessThanK.cpp
// Subarray Product < K — O(n) variable sliding window counting subarrays with product below threshold
#include <vector>

int subarrayProductLessThanK(const std::vector<int>& inputArray, int threshold) {
    if (inputArray.empty() || threshold <= 1) {
        // @step:initialize
        return 0; // @step:initialize
    }

    int leftPointer = 0; // @step:initialize
    int currentProduct = 1;
    int count = 0;

    // Expand the right boundary of the window
    for (int rightPointer = 0; rightPointer < (int)inputArray.size(); rightPointer++) {
        currentProduct *= inputArray[rightPointer]; // @step:expand-window

        // Shrink from the left while product meets or exceeds threshold
        while (currentProduct >= threshold && leftPointer <= rightPointer) { // @step:compare
            currentProduct /= inputArray[leftPointer]; // @step:shrink-window
            leftPointer++; // @step:shrink-window
        }

        // Every subarray ending at rightPointer and starting anywhere in [leftPointer, rightPointer]
        count += rightPointer - leftPointer + 1; // @step:compare
    }

    return count; // @step:complete
}
